spring-cloud-dataflow-completion/src/test/java/org/springframework/cloud/dataflow/completion/TaskCompletionProviderTests.java

```diff
@@ -0,0 +1,246 @@
+		System.out.println(completionProvider.complete("basic", 1));
+		assertThat(completionProvider.complete("basic ", 1), hasItems(
+				proposalThat(is("basic --expression=")),
+				proposalThat(is("basic --expresso="))
+		));
+		// Same as above, no final space
+		assertThat(completionProvider.complete("basic", 1), hasItems(
+				proposalThat(is("basic --expression=")),
+				proposalThat(is("basic --expresso="))
+		));
+	}
+
+	@Test
+	// file | filter -<TAB> => file | filter --foo,etc
+	public void testOneDashShouldReturnTwoDashes() {
+		System.out.println(completionProvider.complete("basic -", 1));
+		assertThat(completionProvider.complete("basic -", 1), hasItems(
+				proposalThat(is("basic --expression=")),
+				proposalThat(is("basic --expresso="))
+		));
+	}
+
+	@Test
+	// basic --<TAB> => basic --foo,etc
+	public void testTwoDashesShouldReturnOptions() {
+		assertThat(completionProvider.complete("basic --", 1), hasItems(
+				proposalThat(is("basic --expression=")),
+				proposalThat(is("basic --expresso="))
+		));
+	}
+
+	@Test
+	// file --p<TAB> => file --preventDuplicates=, file --pattern=
+	public void testUnfinishedOptionNameShouldComplete() {
+		assertThat(completionProvider.complete("basic --foo", 1), hasItems(
+				proposalThat(is("basic --fooble="))
+		));
+	}
+
+	@Test
+	// file | counter --name=<TAB> => nothing
+	public void testInGenericOptionValueCantProposeAnything() {
+		assertThat(completionProvider.complete("http --port=", 1), empty());
+	}
+
+	@Test
+	// plum --use-ssl=<TAB> => propose true|false
+	public void testValueHintForBooleans() {
+		assertThat(completionProvider.complete("plum --use-ssl=", 1), hasItems(
+				proposalThat(is("plum --use-ssl=true")),
+				proposalThat(is("plum --use-ssl=false"))
+		));
+	}
+
+	@Test
+	// basic --enum-value=<TAB> => propose enum values
+	public void testValueHintForEnums() {
+		assertThat(completionProvider.complete("basic --expresso=", 1), hasItems(
+				proposalThat(is("basic --expresso=SINGLE")),
+				proposalThat(is("basic --expresso=DOUBLE"))
+		));
+	}
+
+	@Test
+	public void testUnrecognizedPrefixesDontBlowUp() {
+		assertThat(completionProvider.complete("foo", 1), empty());
+		assertThat(completionProvider.complete("foo --", 1), empty());
+		assertThat(completionProvider.complete("http --notavalidoption", 1), empty());
+		assertThat(completionProvider.complete("http --notavalidoption=", 1), empty());
+		assertThat(completionProvider.complete("foo --some-option", 1), empty());
+		assertThat(completionProvider.complete("foo --some-option=", 1), empty());
+		assertThat(completionProvider.complete("foo --some-option=prefix", 1), empty());
+	}
+
+	/*
+	 * http --use-ssl=tr<TAB> => must be true or false, no need to present "...=tr --other.prop"
+	 */
+	@Test
+	public void testClosedSetValuesShouldBeExclusive() {
+		assertThat(completionProvider.complete("http --use-ssl=tr", 1), not(hasItems(
+				proposalThat(startsWith("http --port=12 --use-ssl=tr "))
```

Comment:
This seems off, the `--port=12` part should come after. May be from stream completion. Will check and fix myself while merging